starknet_py/utils/data_transformer/_calldata_reader.py

```diff
@@ -0,0 +1,40 @@
+from typing import List
+
+CairoData = List[int]
+
+
+class OutOfBoundsError(Exception):
+    def __init__(self, position: int, requested_size: int, remaining_len: int):
+        super().__init__(
+            f"Requested {requested_size} elements, {remaining_len} available."
+        )
+        self.position = position
+        self.requested_size = requested_size
+        self.remaining_len = remaining_len
+
+
+class CalldataReader:
+    _data: List[int]
+    _position: int
+
+    def __init__(self, data: List[int]):
+        self._data = data
+        self._position = 0
+
+    @property
+    def remaining_len(self) -> int:
+        return len(self._data) - self._position
+
+    def consume(self, size: int) -> CairoData:
+        if size < 1:
+            raise ValueError("size must be bigger than 0")
+
+        if size > self.remaining_len:
+            raise OutOfBoundsError(
+                position=self._position,
+                requested_size=size,
+                remaining_len=self.remaining_len,
+            )
+        data = self._data[self._position : self._position + size]
+        self._position += size
+        return data
```

Comment:
If the ability to request `size` consecutive elements is necessary, we could probably have some kind of a wrapper over `itertools.islice` to go along with it being a iterator

Reply:
Requesting n elements is the crucial part. This way you can inform user what went wrong during parsing. I get no benefit from making it iterable, it will only be used to consume batches of known size.

Reply:
Okay, if it's critical like that, let's not make it iterable. Perhaps we could consider renaming `consume` to `read` if the class is named `CalldataReader`.